{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import textwrap\n",
    "import matplotlib\n",
    "import warnings\n",
    "import sys\n",
    "import os\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(current_dir))\n",
    "from SpaMV.utils import compute_gene_topic_correlations\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib.backends.backend_agg import FigureCanvasAgg as FigureCanvas \n",
    "from matplotlib.patches import Patch\n",
    "\n",
    "font = {'size': 7}\n",
    "plt.rc('font', size=8)\n",
    "plt.rc('axes', titlesize=10)\n",
    "plt.rc('axes', labelsize=8)\n",
    "plt.rc('xtick', labelsize=6)\n",
    "\n",
    "matplotlib.rc('font', **font)\n",
    "dataset = 'ccRCC_Y7_T'\n",
    "# load mouse embryo dataset\n",
    "data_rna = sc.read_h5ad('../Dataset/' + dataset + '/adata_RNA.h5ad')\n",
    "\n",
    "cluster_size = 20\n",
    "dr_size = 50\n",
    "width = 13\n",
    "height = 10\n",
    "gt_scale = {'top':1 - 1/height, 'bottom':1 - 4/height, 'left':.05, 'right': .2}\n",
    "clustering_scale = {'top':1 - 1/height, 'bottom':1 - 4/height, 'left': .22, 'right': .72}\n",
    "score_scale = {'top':1 - 1/height, 'bottom':1 - 4/height, 'left': .72, 'right': .82}\n",
    "annotation_scale = {'top': 1-5.2/height, 'bottom': 1-9.3/height, 'left':.05, 'right': .15}\n",
    "dr_scale = {'top': 1-5/height, 'bottom': 1-9.5/height, 'left': .27, 'right':.7}\n",
    "topic_colorbar_scale = {'top': 1-(7.75-.25)/height, 'bottom': 1-(7.75+.25)/height, 'left':.65, 'right': .658}\n",
    "# feature_colorbar_scale = {'top': 1-(5+9.2/3-.2)/height, 'bottom': 1-(5+9.2/3+.2)/height, 'left':.65, 'right': .658}\n",
    "correlation_scale = {'top': 1-5/height, 'bottom': 1-9/height, 'left': .78, 'right': .92}\n",
    "# pathway_scale = {'top': 1-10/height, 'bottom': 1-13/height}\n",
    "\n",
    "fig = plt.figure(figsize=(width, height), dpi=500)\n",
    "###############################################################################################################################\n",
    "# plot clustering results\n",
    "spec_gt = GridSpec(1, 1)\n",
    "spec_gt.update(**gt_scale)\n",
    "he = plt.subplot(spec_gt[0, 0])\n",
    "sc.pl.spatial(data_rna, ax=he, title='H&E', show=False, frameon=False)\n",
    "he.get_xaxis().set_visible(False)\n",
    "he.get_yaxis().set_visible(False)\n",
    "\n",
    "spec_clustering = GridSpec(2, 4, wspace=0.1)\n",
    "spec_clustering.update(**clustering_scale)\n",
    "\n",
    "f_dict = {}\n",
    "f_dict['SpatialGlue'] = plt.subplot(spec_clustering[0, 0])\n",
    "f_dict['CellCharter'] = plt.subplot(spec_clustering[0, 1])\n",
    "f_dict[\"COSMOS\"] = plt.subplot(spec_clustering[0, 2])\n",
    "f_dict['SMOPCA'] = plt.subplot(spec_clustering[1, 0])\n",
    "f_dict['MISO'] = plt.subplot(spec_clustering[1, 1])\n",
    "f_dict['SpaMV'] = plt.subplot(spec_clustering[1, 2])\n",
    "legend = plt.subplot(spec_clustering[:, 3])\n",
    "\n",
    "data_methods = {}\n",
    "for method in ['SpaMV', 'COSMOS', 'CellCharter', 'SpatialGlue', 'MISO', 'SMOPCA']:\n",
    "    data_methods[method] = sc.read_h5ad('../Results/' + dataset + '/' + method + '.h5ad')\n",
    "    data_methods[method] = data_methods[method][data_rna.obs_names.intersection(data_methods[method].obs_names), :]\n",
    "    data_methods[method].obsm['spatial'] = data_rna[data_methods[method].obs_names].obsm['spatial']\n",
    "    data_methods[method].uns['spatial'] = data_rna[data_methods[method].obs_names].uns['spatial']\n",
    "    sc.pl.spatial(data_methods[method], color=method, ax=f_dict[method], show=False)\n",
    "\n",
    "handles, labels = f_dict['SpaMV'].get_legend_handles_labels()\n",
    "legend.legend(handles, labels, loc='center left', frameon=False, ncol=1)\n",
    "legend.axis('off')\n",
    "for f in f_dict.values():\n",
    "    f.get_legend().remove()\n",
    "    f.axes.get_xaxis().set_visible(False)\n",
    "    f.axes.get_yaxis().set_visible(False)\n",
    "\n",
    "###############################################################################################################################\n",
    "# plot clutering evaluation results\n",
    "# plot unsupervised results\n",
    "spec_score = GridSpec(2, 1, hspace=.8)\n",
    "spec_score.update(**score_scale)\n",
    "evaluation_spamv = pd.read_csv('../Results/' + dataset + '/Evaluation_SpaMV.csv')\n",
    "evaluation_miso = pd.read_csv('../Results/' + dataset + '/Evaluation_MISO.csv')\n",
    "evaluation_cosmos = pd.read_csv('../Results/' + dataset + '/Evaluation_COSMOS.csv')\n",
    "evaluation_spatialglue = pd.read_csv('../Results/' + dataset + '/Evaluation_SpatialGlue.csv')\n",
    "evaluation_cellcharter = pd.read_csv('../Results/' + dataset + '/Evaluation_CellCharter.csv')\n",
    "evaluation_smopca = pd.read_csv('../Results/' + dataset + '/Evaluation_SMOPCA.csv', index_col=0)\n",
    "\n",
    "unsupervised = plt.subplot(spec_score[0, 0])\n",
    "melted_df = evaluation_spamv\n",
    "for df in [evaluation_miso, evaluation_cosmos, evaluation_spatialglue, evaluation_cellcharter, evaluation_smopca]:\n",
    "    melted_df = pd.merge(melted_df, df, how='outer')\n",
    "\n",
    "melted_df = melted_df[melted_df['Dataset'] == dataset]\n",
    "metrics = ['jaccard 1', 'jaccard 2']  # your metrics\n",
    "methods = ['MISO', 'CellCharter', 'COSMOS', 'SMOPCA', 'SpatialGlue', 'SpaMV', 'SpaMV (Shared)', 'SpaMV (Transcriptomics related)', 'SpaMV (Metabolomics related)']\n",
    "\n",
    "melted_df = melted_df.drop(['Dataset'], axis=1)\n",
    "# Calculate mean and standard deviation for each algorithm and metric\n",
    "df_means = melted_df.groupby('method').mean().reset_index()\n",
    "df_errors = melted_df.groupby('method').std().reset_index()\n",
    "\n",
    "# Reshape the DataFrames for plotting\n",
    "df_means_melted = df_means.melt(id_vars='method', var_name='Metric', value_name='Mean')\n",
    "df_errors_melted = df_errors.melt(id_vars='method', var_name='Metric', value_name='Std')\n",
    "\n",
    "# Merge the dataframes to include errors\n",
    "df_combined = pd.merge(df_means_melted, df_errors_melted, on=['method', 'Metric'])\n",
    "\n",
    "x = np.arange(len(metrics))  # the label locations\n",
    "width = 0.08  # the width of the bars\n",
    "cat_width=.8\n",
    "colors = {'MISO': '#1f77b4', 'COSMOS': '#ff7f0e', 'spaMultiVAE': '#8c564b', 'CellCharter': '#2ca02c', 'SMOPCA': '#9467bd', 'SpatialGlue': '#bcbd22', 'SpaMV': '#d62728'}\n",
    "ms = ['MISO', 'CellCharter', 'COSMOS', 'SMOPCA', 'SpatialGlue', 'SpaMV']\n",
    "values = {}\n",
    "bar_width = 0.64 \n",
    "for o in ['H3K27ac', 'H3K27me3']:\n",
    "    values[o] = []\n",
    "    for method in ['MISO', 'CellCharter', 'COSMOS', 'SMOPCA', 'SpatialGlue']:\n",
    "        if o == 'H3K27ac':\n",
    "            values[o].append(melted_df[melted_df['method'] == method]['jaccard 1'].mean())\n",
    "        else:\n",
    "            values[o].append(melted_df[melted_df['method'] == method]['jaccard 2'].mean())\n",
    "    if o == 'H3K27ac':\n",
    "        values[o].append(melted_df[melted_df['method'] == 'SpaMV (H3K27ac related)']['jaccard 1'].mean())\n",
    "    else:\n",
    "        values[o].append(melted_df[melted_df['method'] == 'SpaMV (H3K27me3 related)']['jaccard 2'].mean())\n",
    "\n",
    "bottom = np.zeros(len(ms))\n",
    "\n",
    "for boolean, weight_count in values.items():\n",
    "    p = unsupervised.bar(ms, weight_count, bar_width, label=boolean, bottom=bottom)\n",
    "    bottom += weight_count\n",
    "# Add labels and title\n",
    "unsupervised.set_ylabel('Jaccard Similarity')\n",
    "unsupervised.legend(bbox_to_anchor=(1.05, 0.5), loc='center left', borderaxespad=0, frameon=False)\n",
    "unsupervised.axes.set_xticklabels(labels=ms, rotation=45, ha='right', rotation_mode='anchor')\n",
    "unsupervised.axes.xaxis.set_tick_params(pad=0)\n",
    "unsupervised.set_ylim([0, .6])\n",
    "# unsupervised.set_ylim([0, ymax])\n",
    "\n",
    "# Remove the top and right spines\n",
    "unsupervised.spines['top'].set_visible(False)\n",
    "unsupervised.spines['right'].set_visible(False)\n",
    "\n",
    "# Remove the x-axis label\n",
    "unsupervised.set_xlabel('')  # Set the x-axis label to an empty string\n",
    "\n",
    "\n",
    "# plot unsupervised results\n",
    "unsupervised_spamv = plt.subplot(spec_score[1, 0])\n",
    "melted_df = evaluation_spamv\n",
    "for df in [evaluation_miso, evaluation_cosmos, evaluation_spatialglue, evaluation_cellcharter, evaluation_smopca]:\n",
    "    melted_df = pd.merge(melted_df, df, how='outer')\n",
    "\n",
    "melted_df = melted_df[melted_df['Dataset'] == dataset]\n",
    "melted_df = melted_df.drop(['Dataset'], axis=1)\n",
    "methods = ['SpaMV', 'SpaMV (Shared)', 'SpaMV (Transcriptomics related)', 'SpaMV (Metabolomics related)']\n",
    "mean_scores = melted_df.groupby('method').mean().reset_index()\n",
    "std_scores = melted_df.groupby('method').std().reset_index()\n",
    "\n",
    "# Reshape the DataFrames for plotting\n",
    "df_means_melted = mean_scores.melt(id_vars='method', var_name='Metric', value_name='Mean')\n",
    "df_errors_melted = std_scores.melt(id_vars='method', var_name='Metric', value_name='Std')\n",
    "\n",
    "# Merge the dataframes to include errors\n",
    "df_combined = pd.merge(df_means_melted, df_errors_melted, on=['method', 'Metric'])\n",
    "\n",
    "x = np.arange(2)  # the label locations\n",
    "\n",
    "cat_width = .6\n",
    "width = cat_width / 4 * .8 # the width of the bars\n",
    "colors = {'MISO': '#1f77b4', 'COSMOS': '#ff7f0e', 'spaMultiVAE': '#8c564b', 'CellCharter': '#2ca02c', 'SMOPCA': '#9467bd', 'SpatialGlue': '#bcbd22', 'SpaMV': '#d62728'}\n",
    "# Plot bars for each method\n",
    "for i, method in enumerate(methods):\n",
    "    offset = cat_width * i / 4 - (.65-width*4)/.8\n",
    "    algo_data = df_combined[df_combined['method'] == method]\n",
    "    if method == 'SpaMV (Transcriptomics related)':\n",
    "        label = 'Shared + Transcriptome private'\n",
    "    elif method == 'SpaMV (Metabolomics related)':\n",
    "        label = 'Shared + Metabolome private'\n",
    "    elif method == 'SpaMV (Shared)':\n",
    "        label = \"Shared\"\n",
    "    else:\n",
    "        label = 'All'\n",
    "    # unsupervised.bar(x + offset, mean_scores.loc[method], width, label=label, color=color, alpha=alpha, hatch=hatch)\n",
    "    unsupervised_spamv.bar(x + offset, algo_data['Mean'], yerr=algo_data['Std'], width=width, capsize=2, label=label)\n",
    "\n",
    "# Customize the plot\n",
    "unsupervised_spamv.set_ylabel('Jaccard Similarity')\n",
    "unsupervised_spamv.set_xticks(x)\n",
    "unsupervised_spamv.set_xticklabels(['Transcriptome', 'Metabolome'])\n",
    "unsupervised_spamv.set_ylim([0, .4])\n",
    "unsupervised_spamv.legend()\n",
    "# Remove spines\n",
    "unsupervised_spamv.spines['right'].set_visible(False)\n",
    "unsupervised_spamv.spines['top'].set_visible(False)\n",
    "unsupervised_spamv.legend(bbox_to_anchor=(1.02, 0.5), loc='center left', frameon=False) \n",
    "\n",
    "###############################################################################################################################\n",
    "# plot annotated result\n",
    "spec_annotation = GridSpec(1, 1)\n",
    "spec_annotation.update(**annotation_scale)\n",
    "\n",
    "annotated = plt.subplot(spec_annotation[0, 0])\n",
    "spamv_annotated = sc.read_h5ad('../Results/' + dataset + '/SpaMV.h5ad')\n",
    "spamv_annotated.uns = data_rna.uns\n",
    "spamv_annotated.obs['SpaMV_annotation'] = pd.Categorical(spamv_annotated.obs['SpaMV_annotation'], categories=['Stromal', 'Malignant', 'Immune', 'Endothelial'], ordered=True)\n",
    "sc.pl.spatial(spamv_annotated, color='SpaMV_annotation', ax=annotated, frameon=False, title='SpaMV\\nAnnotation', show=False)\n",
    "\n",
    "###############################################################################################################################\n",
    "# plot dimension reduction result\n",
    "spec_dr = GridSpec(4, 6, hspace=.3)\n",
    "spec_dr.update(**dr_scale)\n",
    "z = pd.read_csv('../Results/' + dataset + '/SpaMV_z.csv', index_col=0)\n",
    "col_dict = {}\n",
    "si = 1\n",
    "oi = 1\n",
    "ti = 1\n",
    "for topic in z.columns:\n",
    "    if 'Shared' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(si)\n",
    "        si += 1\n",
    "    elif 'Transcriptomics' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(oi)\n",
    "        oi += 1\n",
    "    else:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(ti)\n",
    "        ti += 1\n",
    "z = z.rename(columns=col_dict)\n",
    "def softmax(row):\n",
    "    shifted = row - np.max(row)  # Prevent overflow\n",
    "    exp_values = np.exp(shifted)\n",
    "    return exp_values / exp_values.sum()\n",
    "z = z.apply(softmax, axis=1)\n",
    "w = [pd.read_csv('../Results/' + dataset + '/SpaMV_w_Transcriptomics.csv', index_col=0), pd.read_csv('../Results/' + dataset + '/SpaMV_w_Metabolomics.csv', index_col=0)]\n",
    "w[0] = w[0].rename(columns=col_dict)\n",
    "w[1] = w[1].rename(columns=col_dict)\n",
    "rna = sc.read_h5ad('../Results/' + dataset + '/adata_RNA_preprocessed.h5ad')\n",
    "met = sc.read_h5ad('../Results/' + dataset + '/adata_MET_preprocessed.h5ad')\n",
    "w[1].index = met.var_names\n",
    "spamv_annotated.obs[z.columns] = z.values\n",
    "spamv_annotated.obs['SpaMV_interpretable'] = z.idxmax(1).values\n",
    "contingency_table = pd.crosstab(spamv_annotated.obs['SpaMV_annotation'], spamv_annotated.obs['SpaMV_interpretable'])\n",
    "\n",
    "# Initialize a dictionary to store results\n",
    "matching = {}\n",
    "\n",
    "# Iterate over each column in the contingency table\n",
    "for column in w[0].columns:\n",
    "    # Find the index with the maximum count for the current column\n",
    "    max_index = contingency_table[column].idxmax()\n",
    "    \n",
    "    # Add the result to the dictionary\n",
    "    if max_index in matching:\n",
    "        matching[max_index].append(column)\n",
    "    else:\n",
    "        matching[max_index] = [column]\n",
    "for column in w[1].columns[-5:]:\n",
    "    # Find the index with the maximum count for the current column\n",
    "    max_index = contingency_table[column].idxmax()\n",
    "    \n",
    "    # Add the result to the dictionary\n",
    "    if max_index in matching:\n",
    "        matching[max_index].append(column)\n",
    "    else:\n",
    "        matching[max_index] = [column]\n",
    "f_dict = {}\n",
    "img_alpha=.5\n",
    "i = 0\n",
    "j = 0\n",
    "# Add a rectangle as a custom background\n",
    "\n",
    "size = 23\n",
    "for cluster in spamv_annotated.obs['SpaMV_annotation'].cat.categories:\n",
    "    for topic in matching[cluster]:\n",
    "        if 'Shared' in topic:\n",
    "            label = topic\n",
    "        elif topic.split(maxsplit=1)[0] == 'Transcriptomics':\n",
    "            label = 'Transcriptome\\n' + topic.split(maxsplit=1)[1]\n",
    "        else:\n",
    "            label = 'Metabolome\\n' + topic.split(maxsplit=1)[1]\n",
    "        f_dict[topic] = plt.subplot(spec_dr[i, j])\n",
    "        if j == 0:\n",
    "            sc.pl.spatial(spamv_annotated, color=topic, ax=f_dict[topic], alpha_img=img_alpha, frameon=True, show=False, title='', colorbar_loc=None, vmax='p99')\n",
    "            # f_dict[topic].set_ylabel(label, fontsize=7)\n",
    "            f_dict[topic].set_title(label, fontsize=7)\n",
    "            if i == 0:\n",
    "                color = '#1F77B4'\n",
    "                ylabel = ' ' * round((size - len(cluster)) / 2) + cluster + ' ' * round((size - len(cluster)) / 2)\n",
    "            elif i == 1:\n",
    "                color = '#FF7F0E'\n",
    "                ylabel = ' ' * round((size - len(cluster)) / 2) + cluster + ' ' * round((size - len(cluster)) / 2)\n",
    "            elif i == 2:\n",
    "                color = '#279E68'\n",
    "                ylabel = ' ' * round((size - len(cluster)) / 2) + cluster + ' ' * round((size - len(cluster)) / 2)\n",
    "            else:\n",
    "                color = '#d62728'\n",
    "                ylabel = ' ' * round((size - len(cluster)) / 2) + cluster + ' ' * round((size - len(cluster)) / 2)\n",
    "            f_dict[topic].set_ylabel(ylabel, bbox=dict(facecolor='none', alpha=0.8, edgecolor=color), labelpad=10, loc='center')\n",
    "            f_dict[topic].spines['top'].set_visible(False)\n",
    "            f_dict[topic].spines['right'].set_visible(False)\n",
    "            f_dict[topic].spines['left'].set_visible(False)\n",
    "            f_dict[topic].spines['bottom'].set_visible(False)\n",
    "            f_dict[topic].set_xlabel('')\n",
    "        else:\n",
    "            sc.pl.spatial(spamv_annotated, color=topic, ax=f_dict[topic], alpha_img=img_alpha, frameon=False, show=False, vmax='p99')\n",
    "            f_dict[topic].set_title(label, fontsize=7)\n",
    "\n",
    "            topic_cb = f_dict[topic].collections[0]\n",
    "            f_dict[topic].collections[0].colorbar.remove()\n",
    "        j += 1\n",
    "    i+=1\n",
    "    j=0\n",
    "\n",
    "spec_topic_colorbar = GridSpec(1, 1)\n",
    "spec_topic_colorbar.update(**topic_colorbar_scale)\n",
    "topic_colorbar = plt.subplot(spec_topic_colorbar[0, 0])\n",
    "tcb = plt.colorbar(topic_cb, cax=topic_colorbar)\n",
    "vmin, vmax = z[topic].min() + .003, z[topic].max() - .02\n",
    "# Define tick positions including intermediate ticks\n",
    "num_ticks = 4  # Total number of ticks\n",
    "ticks = np.linspace(vmin, vmax, num=num_ticks)\n",
    "\n",
    "# Set the ticks on the colorbar\n",
    "tcb.outline.set_visible(False)\n",
    "tcb.set_ticks(ticks)\n",
    "\n",
    "# Create labels where only the first and last are labeled\n",
    "labels = ['Low'] + [''] * (num_ticks - 2) + ['High']\n",
    "\n",
    "# Set these labels on the colorbar\n",
    "tcb.ax.set_yticklabels(labels)\n",
    "tcb.ax.tick_params(size=0)\n",
    "tcb.ax.set_title('Topic\\nAbundance', fontsize=7, loc='left')  # Set the label position to the top\n",
    "###############################################################################################################################\n",
    "# plot correlation result\n",
    "spec_correlation = GridSpec(1, 1)\n",
    "spec_correlation.update(**correlation_scale)\n",
    "dr_score = plt.subplot(spec_correlation[0, 0])\n",
    "def plot_topic_correlation_ratio_multimodal(data, omics_names, z, ax, k_values=None):\n",
    "    \"\"\"\n",
    "    Plot the log2 fold change of mean top-k correlations between modalities for each topic.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    data_list : list\n",
    "        List of AnnData objects, one for each modality\n",
    "    omics_names : list\n",
    "        List of strings containing names of each modality\n",
    "    z : pandas.DataFrame\n",
    "        Topic matrix where rows are cells and columns are topics\n",
    "    k_values : dict or None\n",
    "        Dictionary mapping modality names to their k values for top correlations\n",
    "        If None, defaults to k=20 for RNA and k=5 for others\n",
    "    figsize : tuple\n",
    "        Figure size (width, height)\n",
    "    \"\"\"\n",
    "    # Validate inputs\n",
    "    if len(data) != len(omics_names):\n",
    "        raise ValueError(\"Number of datasets must match number of omics names\")\n",
    "    if len(data) != 2:\n",
    "        raise ValueError(\"This function currently supports exactly 2 modalities\")\n",
    "\n",
    "    # Set default k values if not provided\n",
    "    if k_values is None:\n",
    "        k_values = {name: 5 if name in ['Proteomics'] else 10 for name in omics_names}\n",
    "\n",
    "    for topic in z.columns:\n",
    "        if 'Transcriptomics' in topic:\n",
    "            z = z.rename(columns={topic: 'Transcriptome ' + topic.split(maxsplit=1)[1]})\n",
    "        elif 'Metabolomics' in topic:\n",
    "            z = z.rename(columns={topic: 'Metabolome ' + topic.split(maxsplit=1)[1]})\n",
    "    # Compute correlations for each modality\n",
    "    modality_corrs = {}\n",
    "    for data, name in zip(data, omics_names):\n",
    "        corr_df, _ = compute_gene_topic_correlations(data, z)\n",
    "        modality_corrs[name] = corr_df\n",
    "\n",
    "    # Calculate mean top-k correlations for each modality\n",
    "    topic_means = {name: [] for name in omics_names}\n",
    "\n",
    "    for topic in z.columns:\n",
    "        for name in omics_names:\n",
    "            # Get top k correlations for this modality\n",
    "            top_k_corrs = np.sort(modality_corrs[name][topic].values)[-k_values[name]:]\n",
    "            topic_means[name].append(np.mean(top_k_corrs))\n",
    "\n",
    "    # Calculate log2 fold change\n",
    "    log2_fold_changes = np.log2(np.array(topic_means[omics_names[0]]) /\n",
    "                               np.array(topic_means[omics_names[1]]))\n",
    "\n",
    "    # Create DataFrame with results\n",
    "    result_df = pd.DataFrame({'Topic': z.columns, 'Log2 Fold Change': log2_fold_changes})\n",
    "\n",
    "    # Sort by log2 fold change\n",
    "    result_df = result_df.sort_values('Log2 Fold Change', ascending=True)\n",
    "\n",
    "    # Create horizontal bar plot\n",
    "    bars = ax.barh(range(len(result_df)), result_df['Log2 Fold Change'])\n",
    "\n",
    "    # Color bars based on which modality has stronger correlation\n",
    "    for i, bar in enumerate(bars):\n",
    "        if result_df['Log2 Fold Change'].iloc[i] > 0:\n",
    "            bar.set_color('skyblue')  # First modality stronger\n",
    "        else:\n",
    "            bar.set_color('lightgreen')  # Second modality stronger\n",
    "\n",
    "    # Customize the plot\n",
    "    plt.title(f'Log2 Fold Change of Top Correlations\\n({omics_names[0]} vs {omics_names[1]})')\n",
    "\n",
    "    def split_label(label, threshold=10):\n",
    "       if len(label) <= threshold:\n",
    "           return label\n",
    "\n",
    "       words = label.split()\n",
    "       current_line = \"\"\n",
    "       broken_label = \"\"\n",
    "\n",
    "       for word in words:\n",
    "           if len(current_line) + len(word) + 1 > threshold:\n",
    "               broken_label += current_line.strip() + '\\n'\n",
    "               current_line = word + \" \"\n",
    "           else:\n",
    "               current_line += word + \" \"\n",
    "\n",
    "       broken_label += current_line.strip()  # Add the last line\n",
    "       return broken_label\n",
    "\n",
    "    new_labels = [split_label(label, threshold=16) for label in result_df['Topic']]\n",
    "    # Set topic names as y-axis labels\n",
    "    plt.yticks(range(len(result_df)), new_labels, fontsize=6)\n",
    "\n",
    "    label1 = f'Stronger {omics_names[0]} correlation'\n",
    "    label2 = f'Stronger {omics_names[1]} correlation'\n",
    "\n",
    "    legend_elements = [Patch(facecolor='skyblue', label=label1), Patch(facecolor='lightgreen', label=label2)]\n",
    "    plt.legend(handles=legend_elements, loc='lower right', bbox_to_anchor=(1, -.13), frameon=False)\n",
    "    # Add grid for better readability\n",
    "    plt.grid(axis='x', linestyle='--', alpha=0.5)\n",
    "    plt.tight_layout()\n",
    "    return plt.gcf()\n",
    "\n",
    "plot_topic_correlation_ratio_multimodal([rna, met], ['Transcriptomics', 'Metabolomics'], z, dr_score)\n",
    "\n",
    "plt.tight_layout()\n",
    "fs = 17\n",
    "fig.text(.04, .86, 'a', fontsize=fs, fontweight='bold')\n",
    "fig.text(.2, .92, 'b', fontsize=fs, fontweight='bold')\n",
    "fig.text(.67, .92, 'c', fontsize=fs, fontweight='bold')\n",
    "fig.text(.67, .73, 'd', fontsize=fs, fontweight='bold')\n",
    "fig.text(.04, .37, 'e', fontsize=fs, fontweight='bold')\n",
    "fig.text(.23, .55, 'f', fontsize=fs, fontweight='bold')\n",
    "fig.text(.7, .55, 'g', fontsize=fs, fontweight='bold')\n",
    "fig.text(.04, 0, 'h', fontsize=fs, fontweight='bold')\n",
    "fig.text(.54, 0, 'i', fontsize=fs, fontweight='bold')\n",
    "plt.savefig('../Figures/visualisation_5.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sc.set_figure_params(scanpy=True, fontsize=10)\n",
    "dataset = 'ccRCC_Y7_T'\n",
    "z = pd.read_csv('../Results/' + dataset + '/SpaMV_z.csv', index_col=0)\n",
    "z.index = z.index.astype(str)\n",
    "w = [pd.read_csv('../Results/' + dataset + '/SpaMV_w_Transcriptomics.csv', index_col=0), pd.read_csv('../Results/' + dataset + '/SpaMV_w_Metabolomics.csv', index_col=0)]\n",
    "w[1].index = w[1].index.astype(str)\n",
    "col_dict = {}\n",
    "si = 1\n",
    "oi = 1\n",
    "ti = 1\n",
    "for topic in z.columns:\n",
    "    if 'Shared' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(si)\n",
    "        si += 1\n",
    "    elif 'Transcriptomics' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(oi)\n",
    "        oi += 1\n",
    "    else:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(ti)\n",
    "        ti += 1\n",
    "z = z.rename(columns=col_dict)\n",
    "w[0] = w[0].rename(columns=col_dict)\n",
    "w[1] = w[1].rename(columns=col_dict)\n",
    "data_rna_preprocessed = sc.read_h5ad('../Results/' + dataset + '/adata_RNA_preprocessed.h5ad')\n",
    "sc.pp.scale(data_rna_preprocessed)\n",
    "data_rna_preprocessed.obs['SpaMV'] = z.idxmax(1)\n",
    "data_rna_preprocessed = data_rna_preprocessed[data_rna_preprocessed.obs['SpaMV'].isin(w[0].columns)]\n",
    "data_rna_preprocessed.obs['SpaMV'] = data_rna_preprocessed.obs['SpaMV'].replace({'Transcriptomics private topic 1': 'Transcriptome\\n private topic 1', 'Transcriptomics private topic 2': 'Transcriptome\\n private topic 2', 'Transcriptomics private topic 3': 'Transcriptome\\n private topic 3', 'Transcriptomics private topic 4': 'Transcriptome\\n private topic 4', 'Transcriptomics private topic 5': 'Transcriptome\\n private topic 5'})\n",
    "sc.tl.rank_genes_groups(data_rna_preprocessed, 'SpaMV', method='wilcoxon', key_added='wilcoxon')\n",
    "ax = sc.pl.rank_genes_groups_heatmap(data_rna_preprocessed, key=\"wilcoxon\", groupby=\"SpaMV\", show_gene_labels=False, cmap='coolwarm', vmin=-2, vmax=2, show=False)\n",
    "ax['groupby_ax'].set_ylabel('')\n",
    "# ax['groupby_ax'].set_xticklabels(labels=[ax['groupby_ax'].get_yticklabels()[i].get_text() for i in range(len(ax['groupby_ax'].get_yticklabels()))], rotation=45, ha='right', rotation_mode='anchor')\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Figures/visualisation_5_deg.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_met_preprocessed = sc.read_h5ad('../Results/' + dataset + '/adata_MET_preprocessed.h5ad')\n",
    "sc.pp.scale(data_met_preprocessed)\n",
    "data_met_preprocessed.obs['SpaMV'] = z.idxmax(1)\n",
    "data_met_preprocessed = data_met_preprocessed[data_met_preprocessed.obs['SpaMV'].isin(w[1].columns)]\n",
    "data_met_preprocessed.obs['SpaMV'] = data_met_preprocessed.obs['SpaMV'].replace({'Metabolomics private topic 1': 'Metabolome\\n private topic 1', 'Metabolomics private topic 2': 'Metabolome\\n private topic 2', 'Metabolomics private topic 3': 'Metabolome\\n private topic 3', 'Metabolomics private topic 4': 'Metabolome\\n private topic 4', 'Metabolomics private topic 5': 'Metabolome\\n private topic 5'})\n",
    "\n",
    "sc.tl.rank_genes_groups(data_met_preprocessed, 'SpaMV', method='wilcoxon', key_added='wilcoxon')\n",
    "ax = sc.pl.rank_genes_groups_heatmap(data_met_preprocessed, key=\"wilcoxon\", groupby=\"SpaMV\", show_gene_labels=False, cmap='coolwarm', vmin=-2, vmax=2, show=False)\n",
    "ax['groupby_ax'].set_ylabel('')\n",
    "plt.tight_layout()\n",
    "plt.savefig('../Figures/visualisation_5_dem.pdf', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.gridspec import GridSpec\n",
    "dataset = 'ccRCC_Y7_T'\n",
    "plt.rcParams['font.size'] = 5\n",
    "# sc.set_figure_params(scanpy=True, fontsize=5)\n",
    "z = pd.read_csv('../Results/' + dataset + '/SpaMV_z.csv', index_col=0)\n",
    "w = pd.read_csv('../Results/' + dataset + '/SpaMV_w_Transcriptomics.csv', index_col=0)\n",
    "col_dict = {}\n",
    "si = 1\n",
    "oi = 1\n",
    "ti = 1\n",
    "for topic in z.columns:\n",
    "    if 'Shared' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(si)\n",
    "        si += 1\n",
    "    elif 'Transcriptomics' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(oi)\n",
    "        oi += 1\n",
    "    else:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(ti)\n",
    "        ti += 1\n",
    "z = z.rename(columns=col_dict)\n",
    "w = w.rename(columns=col_dict)\n",
    "rows = w.shape[1]\n",
    "columns = 11\n",
    "d = sc.read_h5ad('../Results/' + dataset + '/adata_RNA_preprocessed.h5ad')\n",
    "d = d[z.index]\n",
    "d.obs[z.columns] = z.values\n",
    "sf = 2\n",
    "space = 0.2\n",
    "fig = plt.figure(figsize=(columns * sf, rows * sf))\n",
    "spec_dr = GridSpec(rows, columns, hspace=space + .2, wspace=space-.2)\n",
    "spec_dr.update(left=0, right=.98, top=.98, bottom=0.02)\n",
    "f_dict = {}\n",
    "i = 0\n",
    "for i in range(rows):\n",
    "    genes = w.nlargest(columns - 1, w.columns[i]).index\n",
    "    for j in range(columns):\n",
    "        f_dict[w.columns[i] + str(j)] = plt.subplot(spec_dr[i, j])\n",
    "        if j == 0:\n",
    "            sc.pl.spatial(d, color=w.columns[i], ax=f_dict[w.columns[i] + str(j)], show=False, frameon=True, legend_loc='none', vmax='p99')\n",
    "            label = w.columns[i].replace('Transcriptomics', 'Transcriptome')\n",
    "            if 'Transcriptome' in label:\n",
    "                label = label.split(maxsplit=1)[0] + '\\n' + label.split(maxsplit=1)[1]\n",
    "            f_dict[w.columns[i] + str(j)].set_title(label, fontsize=14, pad=5)\n",
    "        else:\n",
    "            sc.pl.spatial(d, color=genes[j - 1], ax=f_dict[w.columns[i] + str(j)], show=False, frameon=True, legend_loc='none', vmax='p99', cmap='coolwarm')\n",
    "            f_dict[w.columns[i] + str(j)].set_title(genes[j - 1], fontsize=14, pad=5)\n",
    "        f_dict[w.columns[i] + str(j)].axes.get_xaxis().set_visible(False)\n",
    "        f_dict[w.columns[i] + str(j)].axes.get_yaxis().set_visible(False)\n",
    "        \n",
    "plt.savefig('../Figures/visualisation_5_11_ccRCC_Y7_T_RNA.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "z = pd.read_csv('../Results/' + dataset + '/SpaMV_z.csv', index_col=0)\n",
    "w = pd.read_csv('../Results/' + dataset + '/SpaMV_w_Metabolomics.csv', index_col=0)\n",
    "col_dict = {}\n",
    "si = 1\n",
    "oi = 1\n",
    "ti = 1\n",
    "for topic in z.columns:\n",
    "    if 'Shared' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(si)\n",
    "        si += 1\n",
    "    elif 'Transcriptomics' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(oi)\n",
    "        oi += 1\n",
    "    else:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(ti)\n",
    "        ti += 1\n",
    "z = z.rename(columns=col_dict)\n",
    "w = w.rename(columns=col_dict)\n",
    "rows = w.shape[1]\n",
    "columns = 11\n",
    "d = sc.read_h5ad('../Results/' + dataset + '/adata_MET_preprocessed.h5ad')\n",
    "w.index = d.var_names\n",
    "# d = sc.read_h5ad('../../Dataset/' + dataset + '/adata_H3K27ac_ATAC.h5ad')\n",
    "d = d[z.index]\n",
    "d.obs[z.columns] = z.values\n",
    "sf = 2\n",
    "space = 0.2\n",
    "fig = plt.figure(figsize=(columns * sf, rows * sf), dpi=200)\n",
    "spec_dr = GridSpec(rows, columns, hspace=space + .2, wspace=space-.2)\n",
    "spec_dr.update(left=0, right=.98, top=.98, bottom=0.02)\n",
    "f_dict = {}\n",
    "i = 0\n",
    "for i in range(rows):\n",
    "    genes = w.nlargest(columns - 1, w.columns[i]).index\n",
    "    for j in range(columns):\n",
    "        f_dict[w.columns[i] + str(j)] = plt.subplot(spec_dr[i, j])\n",
    "        if j == 0:\n",
    "            sc.pl.spatial(d, color=w.columns[i], ax=f_dict[w.columns[i] + str(j)], show=False, frameon=True, legend_loc='none', vmax='p99', legend_fontsize=5)\n",
    "            label = w.columns[i].replace('Metabolomics', 'Metabolome')\n",
    "            if 'Metabolome' in label:\n",
    "                label = label.split(maxsplit=1)[0] + '\\n' + label.split(maxsplit=1)[1]\n",
    "            f_dict[w.columns[i] + str(j)].set_title(label, fontsize=14, pad=5)\n",
    "        else:\n",
    "            sc.pl.spatial(d, color=genes[j - 1], ax=f_dict[w.columns[i] + str(j)], show=False, frameon=True, legend_loc='none', vmax='p99', cmap='coolwarm')\n",
    "            f_dict[w.columns[i] + str(j)].set_title('m/z=' + genes[j - 1][:8], fontsize=14, pad=5)\n",
    "        f_dict[w.columns[i] + str(j)].axes.get_xaxis().set_visible(False)\n",
    "        f_dict[w.columns[i] + str(j)].axes.get_yaxis().set_visible(False)\n",
    "        \n",
    "plt.savefig('../Figures/visualisation_5_11_ccRCC_Y7_T_MET.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "# Get the current working directory\n",
    "current_dir = os.getcwd()\n",
    "# Add the parent directory to sys.path\n",
    "sys.path.insert(0, os.path.dirname(os.path.dirname(current_dir)))\n",
    "from Methods.SpaMV_copy.utils import plot_top_positive_correlations_boxplot\n",
    "\n",
    "dataset = 'ccRCC_Y7_T'\n",
    "z = pd.read_csv('../Results/' + dataset + '/SpaMV_z.csv', index_col=0)\n",
    "col_dict = {}\n",
    "si = 1\n",
    "oi = 1\n",
    "ti = 1\n",
    "for topic in z.columns:\n",
    "    if 'Shared' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(si)\n",
    "        si += 1\n",
    "    elif 'Transcriptomics' in topic:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(oi)\n",
    "        oi += 1\n",
    "    else:\n",
    "        col_dict[topic] = topic.rsplit(' ', 1)[0] + ' ' + str(ti)\n",
    "        ti += 1\n",
    "z = z.rename(columns=col_dict)\n",
    "z.index = z.index.astype(str)\n",
    "for topic in z.columns:\n",
    "    z = z.rename(columns={topic: topic.replace('Transcriptomics', 'Transcriptome')})\n",
    "    z = z.rename(columns={topic: topic.replace('Metabolomics', 'Metabolome')})\n",
    "d = [sc.read_h5ad('../Results/' + dataset + '/adata_RNA_preprocessed.h5ad'), sc.read_h5ad('../Results/' + dataset + '/adata_MET_preprocessed.h5ad')]\n",
    "omics_names = ['Transcriptomics', 'Metabolomics']\n",
    "for i in range(len(d)):\n",
    "    plot_top_positive_correlations_boxplot(d[i], z, omics_name=omics_names[i], figsize=(6, 3))\n",
    "    plt.savefig('../Figures/visualisation_5_11_ccRCC_Y7_T_' + omics_names[i] + '_pcc.pdf')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
